### ISSUE TITLE
KupiaSec - Withdrawal race prior to the liquidation of bad debt

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-08-sentiment-v2-judging#533]
KupiaSec

High

# Withdrawal race prior to the liquidation of bad debt

## Summary

When liquidating a bad debt, the total assets of the base pool are decreased, resulting in a loss of funds for the shareholders. If shareholders competitively withdraw their assets from the base pool by front-running the liquidation, those who delay their withdrawal will incur greater losses.

## Vulnerability Detail

When the `PositionManager.liquidateBadDebt()` function is called, it triggers the invocation of the `Pool.rebalanceBadDebt()` function (see `L461`). Within the `Pool.rebalanceBadDebt()` function, the borrowing data for the position is deleted, and the `totalDepositAssets` of the base pool is reduced (see `L547` and `L548`). Consequently, shareholders of the base pool suffer losses from all borrowed assets of the position. This could trigger a withdrawal race. If shareholders compete to withdraw their assets from the base pool by front-running the liquidation of bad debt, those who delay their withdrawals will face greater losses.

```solidity
PositionManager.sol

    function liquidateBadDebt(address position) external onlyOwner {
        ...

        // clear all debt associated with the given position
        uint256[] memory debtPools = Position(payable(position)).getDebtPools();
        uint256 debtPoolsLength = debtPools.length;
        for (uint256 i; i < debtPoolsLength; ++i) {
461         pool.rebalanceBadDebt(debtPools[i], position);
            Position(payable(position)).repay(debtPools[i], type(uint256).max);
        }
    }

----------------------------

Pool.sol

    function rebalanceBadDebt(uint256 poolId, address position) external {
        ...

547     pool.totalDepositAssets = (totalDepositAssets > borrowAssets) ? totalDepositAssets - borrowAssets : 0;
548     borrowSharesOf[poolId][position] = 0;
    }
```

Let's consider the following scenario:

1. The state of the base pool:

    - `totalDepositAssets` = $3000
    - `totalDepositShares` = 3000
    - Share of user A = 1500, corresponding asset amount = $1500
    - Share of user B = 1500, corresponding asset amount = $1500
2. Alice has a position that has borrowed $300 from the base pool.

3. Alice's position is unhealthy, and the owner intends to call the `liquidateBadDebt()` function for Alice's position. The expected state of the base pool after the liquidation of bad debt would be:

    - `totalDepositAssets` = $3000 - $300 = $2700
    - `totalDepositShares` = 3000
    - Share of user A = 1500, corresponding asset amount = $1350, loss = $150
    - Share of user B = 1500, corresponding asset amount = $1350, loss = $150
4. However, before the liquidation, user A withdraws all of his assets by front-running. The state then changes to:

    - `totalDepositAssets` = $3000 - $1500 = $1500
    - `totalDepositShares` = 1500
    - Share of user B = 1500
5. After the liquidation:

    - `totalDepositAssets` = $1500 - $300 = $1200
    - `totalDepositShares` = 1500
    - Share of user B = 1500, corresponding asset amount = $1200, loss = $300

As a result, user B incurs a loss of $300, which is double the expected amount.

## Impact

Through a withdrawal race just before the liquidation of a bad debt, late shareholders will incur greater losses.

## Code Snippet

https://github.com/sherlock-audit/2024-08-sentiment-v2/blob/main/protocol-v2/src/PositionManager.sol#L446-L464

https://github.com/sherlock-audit/2024-08-sentiment-v2/blob/main/protocol-v2/src/Pool.sol#L528-L549

## Tool used

Manual Review

## Recommendation

It is recommended to implement a control mechanism for withdrawals, such as a cooldown period.

Duplicate of #66
